src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+    def _compute_estimate_contract_events_query(
+        self,
+        query: ContractEventQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.scalar() == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        """
+        Method called by the client to return a query time estimate.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Returns:
+            Optional[int]
+        """
+
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self._estimate_query_clause(query))
+                if not result:
+                    return None
+
+                return self._compute_estimate(query, result)
+
+        except QueryEngineError as err:
+            logger.warning(f"Cannot perform query on cache database: {err}")
+            return None
+
+    @singledispatchmethod
+    def _perform_query_clause(self, query: QueryType) -> Select:
+        """
+        Create the SQLAlchemy select statement to perform the query.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            `sqlalchemy.sql.expression.TextClause`
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @_perform_query_clause.register
+    def _perform_block_clause(self, query: BlockQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(Blocks.number >= query.start_block)
+            .where(Blocks.number <= query.stop_block)
+            .where(Blocks.number % query.step == 0)
+        )
+
+    @_perform_query_clause.register
+    def _perform_transaction_clause(self, query: BlockTransactionQuery) -> Select:
+        return select([Transactions]).where(Transactions.block_hash == query.block_id)
+
+    @_perform_query_clause.register
+    def _perform_contract_event_clause(self, query: ContractEventQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(ContractEvents.block_number >= query.start_block)
+            .where(ContractEvents.block_number <= query.stop_block)
+            .where(ContractEvents.block_number % query.step == 0)
+        )
```

Comment:
It might make sense to reduce indirection here and merge these down since we no longer have a single execution statement to decode the results, but one per query

Reply:
Not entirely following you here.

Reply:
[9aa7663](https://github.com/ApeWorX/ape/pull/680/commits/9aa7663dd31bd92c89d81ef1b8472447b53ce28d)